test/e2e/vmagent_test.go

```diff
@@ -291,3 +302,411 @@ func createExternalNodeCRD(data *TestData, nodeName string, ifName string, ip st
+		var peerLabel map[string]string
+		var cidr *string
+		if matchLabel {
+			peerLabel = map[string]string{
+				externalNodeLabelKey: peerVM.nodeName,
+			}
+		} else {
+			peerIPCIDR := fmt.Sprintf("%s/32", peerVM.ip)
+			cidr = &peerIPCIDR
+		}
+		port := int32(iperfPort)
+		ruleFunc(proto, &port, nil, nil, nil, nil, nil, nil, cidr, nil, nil, peerLabel,
+			nil, nil, nil, nil, ruleAction, "")
+	case ProtocolICMP:
+		peerIPCIDR := fmt.Sprintf("%s/32", nodeIP(0))
+		ruleFunc(ProtocolICMP, nil, nil, nil, &icmpType, &icmpCode, nil, nil, &peerIPCIDR, nil, nil, nil,
+			nil, nil, nil, nil, ruleAction, "")
+	}
+	anpRule := builder.Get()
+
+	anp, err := data.CreateOrUpdateANP(anpRule)
+	assert.Nil(t, err, "Failed to create Antrea NetworkPolicy")
+	assert.Nil(t, data.waitForANPRealized(t, anp.Namespace, anp.Name, policyRealizedTimeout), "Failed to realize Antrea NetworkPolicy")
+	return anp
+}
+
+func createANPWithFQDN(t *testing.T, data *TestData, name string, namespace string, appliedToVM vmInfo, fqdnSettings map[string]*crdv1alpha1.RuleAction) *crdv1alpha1.NetworkPolicy {
+	eeSelector := map[string]string{externalNodeLabelKey: appliedToVM.nodeName}
+	builder := &AntreaNetworkPolicySpecBuilder{}
+	builder = builder.
+		SetName(namespace, name).
+		SetPriority(3.0).
+		SetAppliedToGroup([]ANPAppliedToSpec{{ExternalEntitySelector: eeSelector}})
+	anpRule := builder.Get()
+	i := 0
+	for fqdn, action := range fqdnSettings {
+		ruleName := fmt.Sprintf("name-%d", i)
+		policyPeer := []crdv1alpha1.NetworkPolicyPeer{{FQDN: fqdn}}
+		ports, _ := GenPortsOrProtocols(ProtocolTCP, nil, nil, nil, nil, nil, nil, nil)
+		newRule := crdv1alpha1.Rule{
+			To:     policyPeer,
+			Ports:  ports,
+			Action: action,
+			Name:   ruleName,
+		}
+		anpRule.Spec.Egress = append(anpRule.Spec.Egress, newRule)
+		i += 1
+	}
+
+	anp, err := data.CreateOrUpdateANP(anpRule)
+	require.Nil(t, err, "Failed to create Antrea NetworkPolicy")
+	assert.Nil(t, data.waitForANPRealized(t, anp.Namespace, anp.Name, policyRealizedTimeout), "Failed to realize Antrea NetworkPolicy")
+	return anp
+}
+
+func verifyANPConnectivity(t *testing.T, data *TestData, srcVM vmInfo, dstVM vmInfo, connected bool, proto AntreaPolicyProtocol, ruleAction crdv1alpha1.RuleAction) error {
+	var err error
+	switch proto {
+	case ProtocolTCP:
+		fallthrough
+	case ProtocolUDP:
+		err = runIperfCommandOnVMs(t, data, srcVM, dstVM, connected, proto == ProtocolUDP, ruleAction)
+	case ProtocolICMP:
+		err = runPingCommandOnVM(t, data, dstVM, connected)
+	}
+	return err
+}
+
+func runPingCommandOnVM(t *testing.T, data *TestData, dstVM vmInfo, connected bool) error {
+	dstIP := net.ParseIP(dstVM.ip)
+	cmd := getPingCommand(pingCount, 0, strings.ToLower(linuxOS), &dstIP)
+	cmdStr := strings.Join(cmd, " ")
+	expCount := pingCount
+	if !connected {
+		expCount = 0
+	}
+	expOutput := fmt.Sprintf("%d packets transmitted, %d received", pingCount, expCount)
+	// Use master Node to run ping command.
+	pingClient := nodeName(0)
```

Comment:
I don't get why it uses srcVM for TCP and UDP but hardcode master Node for ICMP, the parameters of `testANPOnExternalNodeWithICMP` is confusing too as it specifies srcVM but actually master VM is used.